fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidKey.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.fluid;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.PacketByteBuf;
+
+import net.fabricmc.fabric.impl.transfer.fluid.FluidKeyImpl;
+
+/**
+ * An immutable association of a fluid and an optional NBT tag.
+ *
+ * <p>Do not extend this class. Use {@link #of(Fluid)} and {@link #of(Fluid, CompoundTag)} to create instances.
+ */
+// TODO: "rendering" this. We need a client-side class that can provide a name (might be server-side as well?), a tooltip, a sprite and a color based on the FluidKey instance.
+@ApiStatus.NonExtendable
+public interface FluidKey {
+	/**
+	 * Retrieve an empty ItemKey.
+	 */
+	static FluidKey empty() {
+		return of(Fluids.EMPTY);
+	}
+
+	/**
+	 * Retrieve an ItemKey with the item and tag of a stack.
+	 */
+	static FluidKey of(Fluid fluid) {
+		return of(fluid, null);
+	}
+
+	/**
+	 * Retrieve an ItemKey with an item and an optional tag.
+	 */
+	static FluidKey of(Fluid fluid, @Nullable CompoundTag tag) {
+		return FluidKeyImpl.of(fluid, tag);
+	}
+
+	/**
+	 * Return true if this key is empty, i.e. its item is Items.AIR, and false
+	 * otherwise.
+	 */
+	boolean isEmpty();
+
+	/**
+	 * Return true if the tag of this key matches the passed tag, and false
+	 * otherwise.
+	 *
+	 * <p>Note: True is returned if both tags are {@code null}.
+	 */
+	boolean tagMatches(@Nullable CompoundTag other);
+
+	/**
+	 * Return true if this key has a tag, false otherwise.
+	 */
+	boolean hasTag();
+
+	/**
+	 * Return the item of this key.
+	 */
```

Comment:
```suggestion
	 */
	 @Nullable
```
so later if we decide to add support for powdered snow, we can

Reply:
I'd say it will have to be a custom `Fluid`. The `Fluid` will be responsible for rendering, so I don't think a `null` fluid is viable (and it would be ambiguous).

Reply:
powdered snow can't be a fluid, since if a mod were to make a fluid placer and place down this fluid-not-fluid it would place the fluid, and not the block. for rendering we can provide a sprite method similar to the sprite registry thing for fluids

Reply:
How would a fluid placer place a potion? Imo the fluid placer would have to use a transaction to convert to a bucket for the placement.

Reply:
it wouldn't, if it's not fluid it wouldn't place it down. Alternatively, we can also just provide a "place" method, and a "canPlace" method.

Reply:
I don't think it's in scope tbh. If we really need to provide fluid placement, we can think about it later.

Reply:
Marking as resolved, we need to focus on the core PR.